streams/shared/src/main/scala/zio/stream/ZSink.scala

```diff
@@ -79,41 +82,72 @@ trait ZSink[-R, +E, +A0, -A, +B] { self =>
    */
   final def <|[R1 <: R, E1, B1 >: B, A00 >: A0, A1 <: A](
     that: ZSink[R1, E1, A00, A1, B1]
-  )(implicit ev: A00 =:= A1, ev2: A1 =:= A00): ZSink[R1, E1, A00, A1, B1] =
+  )(implicit ev: A1 =:= A00): ZSink[R1, E1, A00, A1, B1] =
     (self orElse that).map(_.merge)
 
   private[ZSink] final def ?[A00 >: A0, A1 <: A](implicit ev: A1 =:= A00): ZSink[R, Nothing, A00, A1, Option[B]] =
     new ZSink[R, Nothing, A00, A1, Option[B]] {
-      type State = Option[self.State]
+      import ZSink.internal._
+
+      type State = Optional[self.State, A00, B]
 
-      val initial = self.initial.map(Step.leftMap(_)(Option(_))) orElse
-        IO.succeed(Step.done(None, Chunk.empty))
+      val initial = self.initial.foldM(
+        _ => UIO.succeed(Optional.Error(Chunk.empty)),
+        s =>
+          if (self.cont(s))
+            UIO.succeed(Optional.State(s))
+          else
+            self
+              .extract(s)
+              .fold(
+                _ => Optional.Error(Chunk.empty), { case (b, as) => Optional.Value(b, as) }
+              )
+      )
 
       def step(state: State, a: A1) =
         state match {
-          case None => IO.succeed(Step.done(None, Chunk.single(a)))
-          case Some(state) =>
+          case Optional.State(s1) =>
             self
-              .step(state, a)
-              .fold(
-                _ => Step.done(None, Chunk.single(a)),
-                s => Step.leftMap(s)(Some(_))
+              .step(s1, a)
+              .foldM(
+                _ => UIO.succeed(Optional.Error(Chunk.single(a))),
+                s2 =>
+                  if (self.cont(s2))
+                    UIO.succeed(Optional.State(s2))
+                  else
+                    self
+                      .extract(s2)
+                      .fold(
+                        _ => Optional.Error(Chunk.single(a)), { case (b, as) => Optional.Value(b, as) }
```

Comment:
It's not clear to me that failure to extract means that the leftover should be the last processed element (contrast with a failure to step, in which case it should).

Also, why extract at all during `step`? Because of `cont`? You could add another term to the sum type to signal that the `self` sink has finished. E.g.:
```
case class Done(s: self.State)
```

Reply:
Agreed. Done. Let's wait for CI.